### DISCUSSION TITLE
📦 Use automatic semantic versioning?

### DISCUSSION THREAD

**Discussion by mrjosh** [mrjosh/helm-ls#108] [Ideas]
## Overview
As our project grows, maintaining consistent and meaningful versioning is crucial for both developers and users. 
To automate this process, I propose integrating [semantic-release](https://github.com/semantic-release/semantic-release) into our CI/CD pipeline. 
This tool automatically manages versioning, changelogs, and package publishing based on commit messages following the [Angular Commit Message Conventions](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-format) specification (which we're already following).

## Proposal
Objective: Automatically update the project version, more importantly generate changelogs, and publish releases whenever changes are pushed to the main branch.

## Tools
We will use `semantic-release` along with relevant plugins for GitHub and Git.

## Expected Outcome
By integrating semantic-release, every push to main will:
* Automatically bump the version number according to the changes (major, minor, or patch).
* Update the changelog with relevant details.
* Create a new release on GitHub with the updated version and changelog.

This will ensure consistent version management and reduce the manual effort required for release management.

## Feedback
I welcome any feedback or suggestions on this proposal.
If you have experience with `semantic-release` or see potential issues with this approach, please share your thoughts!

Thanks.

**Comment by mrjosh** [comment]
I really want to know your thoughts on this @qvalentin

**Comment by qvalentin** [comment]
Sounds great. I'm a fan of sematic versioning. I've basically just delayed until we say that the project is no longer in early development. But I would say we've reached that point.

Setting up `semantic-release` sounds like a good idea. Just one thing I would like to bring up: Should there be a release for every commit on main? Currently I'm grouping some commits into one release and also minor changes (like updating the readme) should not cause a release.
I guess this can be configured with `semantic-release`?

I also know [release-please](https://github.com/googleapis/release-please) which is quite similar.
But I have not real preference over a tool, as long as the workflow is fitting.

**Reply by mrjosh** [reply]
Thank you for the response and yeah i think we passed the point of early development too.

We need to have a good releasing plan now
Because, i see that this project just going as an incremental release versioning
***like... we already at version `0.0.22` which i think it's not ideal. ( we can reach version 0.0.99 very soon 😄 )***

Certainly a Good Point you mentioned, what if we change a `README` file or just the `Docs`, i should say, as in the [Angular Commit Message Conventions](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-format) guide says:
```
<type>(<scope>): <short summary>
  │       │             │
  │       │             └─⫸ Summary in present tense. Not capitalized. No period at the end.
  │       │
  │       └─⫸ Commit Scope: animations|bazel|benchpress|common|compiler|compiler-cli|core|
  │                          elements|forms|http|language-service|localize|platform-browser|
  │                          platform-browser-dynamic|platform-server|router|service-worker|
  │                          upgrade|zone.js|packaging|changelog|docs-infra|migrations|
  │                          devtools
  │
  └─⫸ Commit Type: build|ci|docs|feat|fix|perf|refactor|test
```

The type of the commit can be one of the following:

* build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
* ci: Changes to our CI configuration files and scripts (examples: CircleCi, SauceLabs)
* docs: Documentation only changes
* feat: A new feature
* fix: A bug fix
* perf: A code change that improves performance
* refactor: A code change that neither fixes a bug nor adds a feature
* test: Adding missing tests or correcting existing tests

also, In the [Commit Message Format](https://github.com/semantic-release/semantic-release?tab=readme-ov-file#commit-message-format), The table shows which commit message gets you which release type when semantic-release runs (using the default configuration):

| Commit message                                                                                                                                                                                   | Release type                                                                                                    |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------------- |
| `fix(pencil): stop graphite breaking when too much pressure applied`                                                                                                                             | ~~Patch~~ Fix Release                                                                                           |
| `feat(pencil): add 'graphiteWidth' option`                                                                                                                                                       | ~~Minor~~ Feature Release                                                                                       |
| `perf(pencil): remove graphiteWidth option`<br><br>`BREAKING CHANGE: The graphiteWidth option has been removed.`<br>`The default graphite width of 10mm is always used for performance reasons.` | ~~Major~~ Breaking Release <br /> (Note that the `BREAKING CHANGE: ` token must be in the footer of the commit) |

So in this case, if you need to change a README file and ignore a release, you just need to make the commit message header as ```docs(readme): update readme``` or just put no type in the header like ```update readme``` will ignore the release

and for a major release, we just need to put a `BREAKING CHANGE: new version` at the footer of the commit

**Reply by qvalentin** [reply]
Sounds good :D

**Reply by qvalentin** [reply]
Do you think you can find the time to set this up? Im not sure if I am able to do it since, I can not configure secrets for the CI.
